contract/RadicleCloudBase.sol

```diff
@@ -0,0 +1,191 @@
+                // active user
+                expiry[org] = expiry[org] + blocksPerMonth;
+            } else {
+                // new user
+                expiry[org] = uint64(block.number + blocksPerMonth);
+            }
+            emit NewTopUp(org, expiry[org]);
+
+            // create a new head, increase shares and link previous head to current one
+            SubStatus memory newHeadStatus = subStatus[lastCreatedSubStatus];
+            newHeadStatus.shares += blocksPerMonth;
+            //newHeadStatus.previous = lastCreatedSubStatus;
+
+            uint64 currBlock = uint64(block.number);
+            subStatus[lastCreatedSubStatus].next = type(uint64).max; // currBlock;
+            lastCreatedSubStatus = currBlock;
+            subStatus[currBlock] = newHeadStatus;
+        }
+    }
+
+    /// @dev Returns smaller number between 'a' and 'b'.
+    function min(uint64 a, uint64 b) internal returns (uint64) {
+        return a < b ? a : b;
+    }
+
+    /// @dev Returns minimum between last buy, expiry, cancel or offset event
+    function _min_of_four() internal returns(uint64) {
+        return min(min(buys[LAST[BUY]], expires[LAST[EXPIRE]]), min(cancels[LAST[CANCEL]], offsets[LAST[OFFSET]]));
+    }
+
+    /// @dev Calculates the amount operator has earned.
+    /// NOTE: BUY/EXPIRE is ordered by nature of chain, CANCEL/OFFSET needs sorting on insert
+    function _calcRealizedRevenue() internal returns (int128)  {
+        uint128 revenue = 0;
+
+        uint64 last = LAST[PROCESSED];
+        uint64 currBlock = uint64(block.number);
+        uint64 rate = RATE;
+
+        while (last < currBlock) {
+            uint64 next_last = _min_of_four();
+
+            if (next_last == MAXBLOCK) {
+                LAST[PROCESSED] = currBlock;
+                break;
+            }
+
+            if (last == next_last) {
+                while (last == next_last) {
+                    if (last == buys[LAST[BUY]]) {
+                        rate += 1;
+                        LAST[BUY] += 1;
+                    }
+                    if (last == expires[LAST[EXPIRE]]) {
+                        rate -= 1;
+                        LAST[EXPIRE] += 1;
+                    }
+                    if (last == cancels[LAST[CANCEL]]) {
+                        rate -= 1;
+                        LAST[CANCEL] += 1;
+                    }
+                    if (last == offsets[LAST[OFFSET]]) {
+                        rate += 1;
+                        LAST[OFFSET] += 1;
+                    }
+                    next_last = _min_of_four();
+                }
+                continue;
+            }
+            revenue += rate * (min(currBlock, next_last) - last);
+            last = next_last;
+        }
+
+        RATE = rate;
+        return int128(revenue);
+    }
+
+    function updateRealizedRevenue(address destination) public {
+        int128 amount = _calcRealizedRevenue();
+        _withdraw(destination, amount);
+    }
```

Comment:
I'd just have two functions `withdraw(address)` and `withdraw(address,uint)`.

Reply:
Also, I don't see this function actually updating anything, ie. it seems like you can call it as many times as you want, and drain the balance.

Reply:
`_calcRealizedRevenue()` updates `lastProcessed`, calling it again won't give any more revenue.